src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs

```diff
@@ -270,5 +271,132 @@ internal static void AddCustomAttributesToExtensions(this IDictionary<string, IO
+        private static IEdmEntityType EntityTypeFromOperationSegment(this ODataSegment segment)
+        {
+            CheckArgumentNull(segment, nameof(segment));
+
+            if (segment is ODataOperationSegment operationSegment)
+            {
+                IEdmOperationParameter bindingParameter = operationSegment.Operation.Parameters.First();
+                IEdmTypeReference bindingType = bindingParameter.Type;
+
+                if (bindingType.IsCollection())
+                {
+                    bindingType = bindingType.AsCollection().ElementType();
+                }
+
+                return bindingType.AsEntity().EntityDefinition();
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Attempts to add the specified <paramref name="path"/> and <paramref name="pathItem"/> to the <paramref name="pathItems"/> dictionary. 
+        /// </summary>
+        /// <param name="pathItems">The target dictionary.</param>
+        /// <param name="context">The OData context</param>
+        /// <param name="path">The key to be added.</param>
+        /// <param name="pathItem">The value to be added.</param>
+        /// <returns>true when the key and/or value are successfully added/updated to the dictionary; 
+        /// false when the dictionary already contains the specified key, and nothing gets added.</returns>
+        internal static bool TryAddPath(this IDictionary<string, OpenApiPathItem> pathItems,
+            ODataContext context,
+            ODataPath path,
+            OpenApiPathItem pathItem)
+        {
+            CheckArgumentNull(pathItems, nameof(pathItems));
+            CheckArgumentNull(context, nameof(context));
+            CheckArgumentNull(path, nameof(path));
+            CheckArgumentNull(pathItem, nameof(pathItem));
+
+            OpenApiConvertSettings settings = context.Settings.Clone();
+            settings.EnableKeyAsSegment = context.KeyAsSegment;
+
+            string pathName = path.PathTemplate ?? path.GetPathItemName(settings);
+
+            if (!pathItems.TryAdd(pathName, pathItem))
+            {
+                if (path.LastSegment is not ODataOperationSegment lastSegment)
+                {
+                    Debug.WriteLine("Duplicate path: " + pathName);
+                    return false;
+                }
+
+                int secondLastSegmentIndex = 2;
+                if (path.Count < secondLastSegmentIndex)
+                {
+                    Debug.WriteLine($"Invalid path. Operation not bound to any entity. Path: {pathName}");
+                    return false;
+                }
+
+                ODataSegment lastSecondSegment = path.Segments.ElementAt(path.Count - secondLastSegmentIndex);
+                IEdmEntityType boundEntityType = lastSecondSegment.EntityTypeFromPathSegment();
+
+                IEdmEntityType operationEntityType = lastSegment.EntityTypeFromOperationSegment();
```

Comment:
As these two functions can return null, we should probably check for null returns before using the returned values later on.

Reply:
Resolved